atmat/atphysics/LongitudinalDynamics/atBunchLength.m

```diff
@@ -6,17 +6,29 @@
 %
 % Ib is the bunch current [A] (it may be a vector for multiple values)
 % Zn is the longitudinal broadband impedance [Ohms]
-% ring is the at ring without radiation
+% ring is the at lattice (4D or 6D)
+%
 % BL is the bunch length in metres 
 %
-%   see also: BunchLength
+%   see also: BunchLength, blgrowth
 
-rp=ringpara(ring);
-[E0,~,Vrf,h,U0]=atenergy(ring);
-alpha=rp.alphac;
-sigdelta=rp.sigma_E;
+is6d=check_6d(ring);
 circ=findspos(ring,length(ring)+1);
 
-BL = BunchLength(Ib,Zn,Vrf,U0,E0,h,alpha,sigdelta,circ);
-
+if ~is6d
+    [E0,~,Vrf,h,U0]=atenergy(ring);
+    rp=ringpara(ring);
+    alpha=rp.alphac;
+    sigdelta=rp.sigma_E;
+    BL = BunchLength(Ib,Zn,Vrf,U0,E0,h,alpha,sigdelta,circ);
+else
+    [~,ringdata]=atx(ring,1);
+    [~,~,Vrf,h]=atenergy(ring);
+    alpha = ringdata.alpha;
+    E0 = ringdata.energy;
+    U0 = ringdata.eloss;
+    sigdelta = ringdata.espread;
+    bl0 = ringdata.blength;
+    BL = bl0 * blgrowth(Ib,Zn,Vrf,U0,E0,h,alpha,sigdelta);
```

Comment:
There are some possible optimizations here:
alpha -> mcf(disable_6d(ring))
energy, vrf, h -> atGetRingProperties
sigdelta, bl0 -> ohmi_envelope
U0 -> atgetU0(ring, 'method', 'tracking')

Reply:
On top of @swhite2401 proposal:

`is6d` and `alpha` are also available through `atGetRingProperties`. Asking for `is6d`, `Vrf`, `E0`, `h`, `alpha` and `circ` in one call (before the `if` test) minimises the number of passes through the lattice. Be careful to use the "single cell" attributes: `cell_rf_voltage`, `cell_harmnumber`, `cell_length`.

You could also get both `bl0` and `sigdelta` from `ringpara` and completely forget `BunchLength`.